### PR TITLE
Add synchronous ("parasitic") ExecutionContext

### PR DESCRIPTION
**Pull request by viktorklang**
A synchronous, trampolining, ExecutionContext has been used for a long time *within* the Future implementation to run controlled logic as cheaply as possible.

I believe that there is a significant number of use-cases where it makes sense, for efficiency, to execute logic synchronously in a safe(-ish) way without having users to implement the logic for that ExecutionContext themselves—it is tricky to implement to say the least.

It is important to remember that `ExecutionContext` should be supplied via an implicit parameter, so that the caller can decide where logic should be executed. The use of `ExecutionContext.parasitic` means that logic may end up running on Threads/Pools that were not designed or intended to run specified logic. For instance, you may end up running CPU-bound logic on an IO-designed pool or vice versa. So use of `parasitic` is only advisable when it really makes sense. There is also a real risk of hitting StackOverflowErrors for certain patterns of nested invocations where a deep call chain ends up in the `parasitic` executor, leading to even more stack usage in the subsequent execution. Currently the `parasitic` ExecutionContext will allow a nested sequence of invocations at max 16, this may be changed in the future if it is discovered to cause problems.

### CODE REVIEW

**Comment by huntc** [comment]
Seems like a good idea. Context switching can be hell, particularly when there are few resources available and it makes no sense to context switch for just a map etc ie a single core. I think I’d find myself using this in most places.

**Comment by viktorklang** [comment]
@huntc It's important to remember that it is still considered to be The Way To Do It(tm) to have the EC passed in as an implicit parameter so that the caller can decide where something is run, so using the callingThread ExecutionContext directly should be rather rare.

**Comment by huntc** [comment]
Thanks for the reminder - and yes, I'm on the same page.

**Comment by viktorklang** [comment]
@huntc I mainly added that comment to make sure that other readers of this conversations get the right expectations. I am still a bit weary of exposing this feature to the general population, but I think offering a safe alternative is better than DIY solutions for things like these. :)

**Comment by lihaoyi** [comment]
CC @sjrd I vaguely remember we discussed this half a decade ago around the time https://github.com/scala-js/scala-js/issues/2102 happened

**Comment by lihaoyi** [comment]
Also relevant https://grokbase.com/t/gg/scala-user/1421nzzvvb/should-there-be-a-runnow-executioncontext-in-executioncontext-implicits

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (NthPortal) [inline src/library/scala/concurrent/ExecutionContext.scala]: this warning honestly doesn't seem strong enough to me. is there any valid situation where you should call blocking code using this `ExecutionContext`?

Comment (NthPortal) [reply src/library/scala/concurrent/ExecutionContext.scala]: I personally lean towards something like "DO NOT call any blocking code ..."

Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: Great point, @NthPortal. I have updated the Scaladoc—what do you think of the new text?

File: src/library/scala/concurrent/BatchingExecutor.scala
Comment (WellingR) [inline src/library/scala/concurrent/BatchingExecutor.scala]: If you catch an `InterruptedException` shouldn't you interrupt the current thread again?

Comment (viktorklang) [reply src/library/scala/concurrent/BatchingExecutor.scala]: @WellingR I've been back-and-forth on that one, since control is handed back to the "pool" already. Question is when it should be reset—since we cannot *not execute* any of the `Runnable`s since then they will never be executed.

Comment (viktorklang) [reply src/library/scala/concurrent/BatchingExecutor.scala]: @WellingR I just also remembered: If there's an InterruptedException thrown by user logic, they're using this EC to run blocking logic—which is documented to not be done. :/

**Comment by mdedetrich** [comment]
I think it makes sense for making this public, I mean the current situation with `Future` has gotten so ridiculous that akka ended up having to implement a "fast" `Future` to solve this exact problem (i.e. having to use an `ExecutionContext` for small `.map` operations).

I am also wondering whether having control over trampolining or not should be desired for those cases where you need to squeeze even more performance.

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (jroper) [inline src/library/scala/concurrent/ExecutionContext.scala]: I like that the potential performance improvements from this over a thread pool executor are not mentioned - the last thing we want is people seeing this and saying "oh magic performance improvement secret sauce! I'm going to use this!" without reading the rest of the docs. Also, I wonder if we can be even clearer about the need for care. Something like "Symptoms of misusing this `ExecutionContext` include deadlocks and a potential for your application to slow to a crawl under minimal load."

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (retronym) [inline src/library/scala/concurrent/ExecutionContext.scala]: Should we add explicit memory fences so that this execution context establishes the happens-before relationships that the normal one would?

Comment (jroper) [reply src/library/scala/concurrent/ExecutionContext.scala]: I don't think it's necessary.

There's no need for a memory barrier since it's guaranteed that the runnable will run on the calling thread, so there is an intrinsic happens before relationship there. So, as long as any state captured by the runnable is appropriately synchronised before being submitted to the executor (which this executor can't do anything about, memory barrier or not, if it's not), then everything should be fine.

So if we then think about this in the context of the Scala Future API, the `DefaultPromise` implementation uses `AtomicReference.compareAndSet` to query for completion state and/or add callbacks to be executed (eg in `map`) and remove them before executing them again (eg in `complete`). There is a memory barrier here (eg, on x86 its implemented using `CMPXCHG` which gives a memory barrier), so any state held by callbacks that then eventually get submitted to this executor will have their state synchronised so as to guarantee happens before before being submitted to the executor. Other implementations of `Future` likewise must implement a memory barrier on all their callback capturing methods, if they don't then adding a memory in this executor will only save them in a subset of situations, specifically, it couldn't save them, for example, if callbacks captured by `map` didn't have a memory barrier, since by the time this executor gets involved, the calling thread that registered that callback is gone and its memory cannot be synchronised.

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: @jroper Great points, James. I'll craft a message about symptoms of misuse.

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: @retronym @jroper Yes, since the execution does not leave the calling thread then there is no need for any fences.

**Comment by viktorklang** [comment]
@lihaoyi Thanks for reminding me of that 5 year old conversation! :-) I guess I've grown more flexible in my thinking over the years :)

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (viktorklang) [inline src/library/scala/concurrent/ExecutionContext.scala:159]: @jroper @NthPortal Is this an adequate explanation?

**Comment by viktorklang** [comment]
It would seem like there are basically only positive feedback on this proposal, so I have switched it over to Ready for review.

**Comment by sjrd** [comment]
I am a bit worried that this will set wrong expectations, not in terms of performance, but in terms of correctness. Asynchronous code, requesting `implicit ExecutionContext`s, does not always follow the EC for *all* operations. Sometimes, they delegate to other asynchronous APIs, and provide callbacks in which they will complete `Promise`s. Such code will *still* complete asynchronously, even if you call it with a `callingThread` EC. If a user calls such code with a `callingThread` EC and then expects the returned `Future` to be synchronously completed, their code won't work. It's even more annoying that the implementation could subtly change in later versions of a library, and then code that used to be OK by chance will break.

This is something we've experienced in Scala.js with the `runNow` execution context we used to have, and is one of the reasons we deprecated it, and removed it in 1.x.

**Comment by viktorklang** [comment]
@sjrd I think that's a good point. I think the name `runNow` gives a bit different expectations than `callingThread`—but I agree that it should be better spelled out in the documentation for it. Do you have any suggestion as what would be the clearest formulation of such text?

**Comment by sjrd** [comment]
Something like

> Using `callingThread` does not give any guarantee that the resulting `Future[T]` will be completed by the time the callee returns. The caller must still use combinators and/or `onComplete`-like methods to process the result of the `Future` when it becomes ready.

The wording can probably be improved, but the above should convey the message that I think is important.

**Comment by viktorklang** [comment]
@sjrd Ok, thank you! Since ECs are not exclusive to Futures, I was struggling a bit to find the optimal place to add such documentation—but I guess it's fine to add it in the `callingThread` EC as it is most likely that it will be used for such code.

**Comment by NthPortal** [comment]
> It would seem like there are basically only positive feedback on this proposal

I have mixed feelings, but it took me a while to figure out what I wanted to say.

First, there is the concern that where `callingThread` executes is non-deterministic. If you perform some operation on a `Future`, and it is already completed when the operation is performed, that operation will be executed on the current thread; otherwise, the operation will be executed by whatever `ExecutionContext` completes the `Future`. This non-determinism can lead to nasty bugs (as noted by Sébastien and others).

Second, it allows one to perform operations with an `ExecutionContext` they would not otherwise have access to. If the operation performed using `callingThread` is too expensive, it may slow or starve the `ExecutionContext`. While it would be possible to write your own implementation of `callingThread` and access the same inaccessible `ExecutionContext`, making it part of the standard library makes it *significantly* more accessible and more likely to be misused. If people have to implement it themselves, it will be mostly only experts who know when it's safe to use it implementing it; if it's in the standard library, anyone might think it's a good idea to use (when it isn't). This change makes it very easy to starve someone else's `ExecutionContext`.

I am personally more concerned about the second problem, though both are important. I think both problems can be mostly addressed by good documentation, though I also wonder if it is possible or a good idea to make it somehow less accessible (a worse name? nested inside something?). I will try to review the docs more this weekend.

**Comment by viktorklang** [comment]
@NthPortal Yes, I agree with everything you wrote. I think improving the documentation is the right thing to do. It is worth noting that none of the issues you mention is *new* capabilities (users can already write this logic themselves to starve or disrupt other ExecutionContexts). Perhaps with better documentation, and possibly another (more discouraging?) name, we can arrive at a solution which is to be deemed acceptable?

**Comment by tarsa** [comment]
> Perhaps with better documentation, and possibly another (more discouraging?) name, we can arrive at a solution which is to be deemed acceptable?

What about `ExecutionContext.threadless` or even `ExecutionContext.parasitic`? With documentation like:
> This ExecutionContext latches on all threads it is called from. Threads are blocked until their thread local task queues are empty. When this ExecutionContext is used indirectly (through some higher level abstraction like Future) it can be unpredictable to which threads it will latch on. Possibly it can block threads that are critical to application stability and responsiveness.

That description actually explains the mechanics and the problem with careless usage.

**Comment by jroper** [comment]
`threadless` I think will lead to confusion, at worst people might think "using this executor doesn't use a thread and so is better because it uses less resources", which from my experience is more plausible that there are people that might think that than it sounds (like people that think that any blocking operation can be made non blocking by wrapping it in a future).

`parasitic` is too negative, there are plenty of genuinely good use cases for a calling thread executor, they shouldn't have such negative undertones in their code.

**Comment by tarsa** [comment]
> `parasitic` is too negative, there are plenty of genuinely good use cases for a calling thread executor, they shouldn't have such negative undertones in their code.

Yeah, there are plenty of good use cases, but only if you're fully aware of the pitfalls. If you are then you can rename it in your project, e.g.:
```scala
object CustomThreadPools {
  implicit val notAScaryPool: ExecutionContext = ExecutionContext.parasitic
}
```
"parasitic" is still better than "pimp my library" :P

`callingThread` is IMO bad because when used with abstraction like Future in a code like this: ```existingFuture.map(x => x.doSomething)(ExecutionContext.callingThread)``` it's very unintuitive what `callingThread` refers to and users would probably think the "calling thread" is the thread that invoked the `map` operation.

**Comment by SethTisue** [comment]
`sameThread`? `oneThread`? `unthreaded`?

**Comment by WellingR** [comment]
`reUseThread`? `synchronous`? `hitchhike`?

I do like the documentation which @tarsa wrote a few comments above.

**Comment by viktorklang** [comment]
I liked the following:

>
This ExecutionContext latches on all threads it is called from. Threads are blocked until their thread local task queues are empty. When this ExecutionContext is used indirectly (through some higher level abstraction like Future) it can be unpredictable to which threads it will latch on. Possibly it can block threads that are critical to application stability and responsiveness.


W.r.t. naming `parasitic` is the best option I've heard so far.

Another option is `freeloader`. Or `hijacker`.

**Comment by mdedetrich** [comment]
Whats wrong with `currentExecutionThread`?

**Comment by viktorklang** [comment]
@mdedetrich What's the intuition for what `current` means?

**Comment by mdedetrich** [comment]
Oh right, how about `hijackedExecutionThread` then?

**Comment by viktorklang** [comment]
@mdedetrich ExecutionThread seems a bit wordy/redundant. How about `ExecutionContext.threadHijacker`

**Comment by viktorklang** [comment]
Or `ExecutionContext.piggybacking`

**Comment by WellingR** [comment]
I like `piggybacking` or `piggybacked`, it is not quite as negative as `parasitic` (which also is a good option), but it does describe what the execution context does, and the name might trigger a potential user to read the scaladoc which might help this user decide whether it is good idea to use this or not.

**Comment by mdedetrich** [comment]
Ah yes, piggy back (as a native english speaker) sounds like the best phrase.

**Comment by tarsa** [comment]
As a non-native English speaker I vote for `threadHijacker` over `piggybacking` as "hijacking" sounds more familiar to me than "piggybacking". I have an impression that "hijacking" is more often used in context of IT. It's not a strong preference, though.

**Comment by huntc** [comment]
Naming. I personally tahini the original `callingThread` name is fine.

**Comment by tarsa** [comment]
Seems we're hit by https://en.wikipedia.org/wiki/Arrow%27s_impossibility_theorem :]

To repeat: `ExecutionContext.callingThread` gives the wrong intuition in the use case it's designed for, i.e: `existingFuture.map(x => x.doSomething)(ExecutionContext.callingThread)`

**Comment by huntc** [comment]
I see. Perhaps then:

```
existingFuture.map(x => x.doSomething)(ExecutionContext.trampoline)
```

Call it what it is? Confusing enough to avoid. ;-)

**Comment by viktorklang** [comment]
trampoline sounds fun and harmless—we want something which is cautioning.
-- 
Cheers,
√

**Comment by huntc** [comment]
Trampolines are dangerous! :-)

**Comment by viktorklang** [comment]
trapeze? 😂
-- 
Cheers,
√

**Comment by NthPortal** [comment]
originally, `parasitic` seemed to harsh to me, but then it grew on me. it reflects what it does really well (it steals processing power from some thread/EC, which may even be your own), and is only safe if you're stealing very little processing. I'm also okay with `hijacking`, but I like `parasitic` better.

**Comment by viktorklang** [comment]
I’d be ok with parasitic too. Has a name which leads the user to read its
docs.
-- 
Cheers,
√

**Comment by viktorklang** [comment]
I've committed an update to `parasitic`, have a look at it and the revised docs: https://github.com/scala/scala/pull/7784/commits/cc5ae2a2fbda0e1eacd3e398a6f10696e741f1cd

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: @jroper The new working name is `parasitic`—please check out the new docs. :)

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (NthPortal) [inline src/library/scala/concurrent/ExecutionContext.scala]: I'm having trouble understanding why this reads from `Implicits.global` and does a cast, rather than `Implicits.global` reading from this and not needing to cast (not from this PR, ik)

Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: @NthPortal Great question! It's because Implicits.global needs to be typed `ExecutionContext` and not `ExecutionContextExecutor` otherwise its more specific type would take precedence over other implicit `ExecutionContext`s in scope. The non-implicit `ExecutionContext.global` can be typed as `ExecutionContextExecutor` so it can be used for Java APIs requiring `Executor`.

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (NthPortal) [inline src/library/scala/concurrent/ExecutionContext.scala]: Is it necessary to say 'In order to maximize application responsiveness' and that it's 'strongly advised'? What are your thoughts on shortening the sentence to just 'Only execute logic which will quickly return control to the caller'?

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (NthPortal) [reply src/library/scala/concurrent/ExecutionContext.scala]: @viktorklang what I meant was, why not instead do this?

```scala
lazy final val global: ExecutionContextExecutor = impl.ExecutionContextImpl.fromExecutor(null: Executor)
object Implicits {
  implicit final def global: ExecutionContext = ExecutionContext.global // no cast needed!
}
```

**Review by NthPortal — APPROVED** [review]
I like this approach. I think most people, upon seeing `parasitic`, will say "I don't think I want to use that one," which is what we want. There might be some shed painting on the exact phrasing of the docs, but overall looks great to me.

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: @NthPortal  Updated! :)

File: src/library/scala/concurrent/ExecutionContext.scala
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: @NthPortal Done!

**Comment by NthPortal** [comment]
@viktorklang looks like messing with which is the primary definition of `global` changed the phrasing of a partest output

```diff
  t8849.scala:8: error: ambiguous implicit values:
- both lazy value global in object Implicits of type => scala.concurrent.ExecutionContext
+ both method global in object Implicits of type => scala.concurrent.ExecutionContext
```

**Comment by viktorklang** [comment]
@NthPortal Fixed t8849

**Review by WellingR — APPROVED** [review]

**Comment by tarsa** [comment]
Unless I missed something, documentation still doesn't seem to mention unpredictable outcomes when using this ExecutionContext through abstractions, i.e. I mean this bit:
> When this ExecutionContext is used indirectly (through some higher level abstraction like Future) it can be unpredictable to which threads it will latch on. Possibly it can block threads that are critical to application stability and responsiveness.

Wording can be changed to be consistent with the rest of documentation, e.g. "latch on" -> "steal time".

**Comment by viktorklang** [comment]
Piotr, I have updated the docs to include a mention about Futures. Have a
look!
-- 
Cheers,
√

**Comment by tarsa** [comment]
Yep, looks good.
:+1:

File: src/library/scala/concurrent/Future.scala
Comment (NthPortal) [inline src/library/scala/concurrent/Future.scala:873]: I'm just noticing that this behaviour changes slightly - internal exceptions are no longer wrapped in `IllegalStateException`s - is that desired? (I realise we don't want to duplicate logic)

Comment (viktorklang) [reply src/library/scala/concurrent/Future.scala:873]: @NthPortal The reason I removed it is because it is not related to scala.concurrent internal callbacks anymore, and the stacktrace will most likely be non-deterministic to look at. :/

**Comment by lihaoyi** [comment]
Here's a bit of a wild idea: would it be possible to make an ExecutionContext that is smart enough to automatically trampoline operations on the same thread, IFF the operations are given the same execution context to run on (by reference equality), and where the ExecutionContext's differ then going through the normal flow?

That would allow high performance in the common case where a bunch of `map`s and `flatMap`s share an ExecutionContext, while still preserving good behavior in the case where we're handing over our asynchronous workflows between different ExecutionContexts

**Comment by tarsa** [comment]
@lihaoyi 
Documentation says:
> Nested invocations of `execute` will be trampolined to prevent uncontrolled stack space growth.

Isn't that what you want?

Other executors also have batching. I'm not sure how it works in detail, but IIUC a single batch is executed on single thread.

**Comment by viktorklang** [comment]
@lihaoyi Yes, that is implemented in https://github.com/scala/scala/pull/7470 :-)

**Comment by viktorklang** [comment]
Squashed some of the commits here. Will merge after CI OKs it.

**Comment by SethTisue** [comment]
@viktorklang people will find their way here from the release notes, so you may want to revise the PR description to reflect the final state

**Comment by viktorklang** [comment]
@SethTisue Thanks Seth—I've revised the PR description.

**Comment by SethTisue** [comment]
new blog post by @wsargent about `ExecutionContext.global`, `ExecutionContext.parasitic`, and `ExecutionContext.opportunistic`:

https://tersesystems.com/blog/2024/06/20/executioncontext.parasitic-and-friends/

**Pull request merged**
